### DISCUSSION TITLE
Proxyquire or Import Stubbing?

### DISCUSSION THREAD

**Discussion by justinmchase** [denoland/deno#16168] [Q&A] [answered]
In nodejs its possible to intercept the calls to require from outside a module for purposes of stubbing and isolating a module for test, such as with the popular library [`proxyquire`](https://www.npmjs.com/package/proxyquire).

This doesn't seem possible in Deno (or for ESM module importing in general) and I'm curious if there is a possible trick to doing this or what are the general thoughts on module isolation and mocking imports, for testing purposes?

**Comment by zachauten** [comment, marked as answer]
Deno's std lib has a function for stubbing, does [this page from the manual](https://deno.land/std@0.159.0/testing) describe what you're looking for? The stub function's implementation is [here](https://github.com/denoland/deno_std/blob/main/testing/mock.ts#L383).

**Reply by justinmchase** [reply]
Thank you, I'm aware of these but the question is how do you stub an import in another module? That's what proxyquire does, it allows you to intercept a modules call to `require` and hand it a stub instead of doing the actual import. How can you do this in Deno?

For example:
##### a.ja
```js
export function a() {
  return 7;
}
```
##### b.js
```js
import { a } from './a.js';
export function b() {
   return a() + 11;
}
```
##### b.test.js
```js
Deno.test({
  name: 'b',
  fn: async () => {
    const stubImports = {
      './a.js': { a: () => 0 } // how can I stub the entire a.js module?
    }
    const { b } = await import('./b.js', stubImports) // not real a real import signature, but I wish it was
    assert.equal(b(), 11)
  }
})
```

**Reply by ayame113** [reply]
After creating another file with the stub implementation, you can use import-map to stub the module.
https://deno.land/manual@v1.26.1/linking_to_external_code/import_maps#import-maps

```ts
// a_stub.ts
export function a() {
  return 0;
}
```

```jsonc
// import_map_for_test.json
{
  "imports": {
    "./path/to/a.ts": "./path/to/a_stub.ts",
  }
}
```

```
deno test --import-map=./import_map_for_test.json
```

Alternatively you can use the method described in https://deno.land/manual@v1.26.1/testing/mocking.

**Reply by justinmchase** [reply]
Ok but now suppose for one test I need one stub but for another test I need a different stub? I basically need an import map _per test_ but this import map feature only appears to be applied at the process level.

Is there no mechanism for stub / spy / sandbox / resetting modules?

**Reply by justinmchase** [reply]
I see in that link its suggesting I export the internals of a module purely so a spy can replace them... I hope it goes without saying that this seems like a major design flaw in the import system. There's no way anyone is going to be wrapping their imports onto another field and export it purely for testing purposes.

Does the import function not have a hook where I can just implement whatever custom resolver I want? I do believe that in node the require function has this and just adding an async api for this seems pretty straight forward. It would surely solve this problem elegantly and probably a bunch of others.

**Reply by ayame113** [reply]
> Does the import function not have a hook where I can just implement whatever custom resolver I want? I do believe that in node the require function has this and just adding an async api for this seems pretty straight forward. It would surely solve this problem elegantly and probably a bunch of others.

I think this is quite difficult and takes a lot of time. As far as I know, things like custom resolvers are forbidden by the spec, I think.

> Import assertions do not influence how an imported asset is evaluated, and they do not influence the HostResolveImportedModule idempotency requirements. This proposal does. Also see https://github.com/tc39/proposal-import-assertions#follow-up-proposal-evaluator-attributes.
https://github.com/tc39/proposal-import-reflection#qa

-----

As a workaround until something like that is introduced, I often create a stub using an object like this:

```ts
// a.ts
export const a = {
  fn() {
    return 7;
  },
}
```

```ts
// b.ts
import { foo } from './a.ts';
export function b() {
   return a.fn() + 11;
}
```

```ts
import { foo } from "./a.ts";
import { b } from "./b.ts";

Deno.test({
  name: "b",
  fn() {
    const aStub = stub(a, "fn", () => 0);

    try {
      assertEquals(b(), 11);
    } finally {
      aStub.restore();
    }
  }
});
```

I also use [dependency injection](https://en.wikipedia.org/wiki/Dependency_injection) too.

```ts
// a.ts
export function a() {
  return 7;
}
```

```ts
// b.ts
import { foo } from './a.ts';

export function b(a: () => number) {
   return a() + 11;
}
```

```ts
import { b } from "./b.ts";

Deno.test({
  name: "b",
  fn() {
    assertEquals(b(() => 0), 11);
  }
});
```

**Reply by justinmchase** [reply]
I'm gonna mark this as the answer, its not quite what I was looking for but I think that what I'm looking for doesn't exist. It appears there are a couple of ecmascript proposals that may change that somewhat but for now you ran through all of the various patterns and techniques available.

**Comment by justinmchase** [comment]
In addition to @zachauten 's suggestions the only other approach I see so far is here:
https://deno.land/x/fake_imports@v0.6.1

Where it basically pre-resolves all of the imports, loads the files into memory, rewrites all of the imports and generates temporary in memory Blob urls with certain modules overwritten with your mock code and then resolves that instead. It even seems to cleverly generate source maps so all the urls are relative to the original files as well as respecting your actual import maps.

Also, I found this ecmascript proposal which may be taking into consideration such scenarios:
https://github.com/tc39/proposal-import-assertions/issues/46
